core/src/instance.rs

```diff
@@ -116,16 +123,32 @@ impl Instance {
     ) -> Vec<Observer> {
         // Mutate state
         {
+            let new_state: State;
+
+            {
+                // Only get a read lock first so code in reducers can read state as well
+                let state = self
+                    .state
+                    .read()
+                    .expect("owners of the state RwLock shouldn't panic");
+
+                // Create new state by reducing the action on old state
+                new_state = state.reduce(
+                    context.clone(),
```

Comment:
This is kind of weird since `context` also holds a state...
maybe `reduce` doesn't need to be a method anymore, and is better off as a function that gets current state from `context` argument?